iothub_client/inc/iothub_device_client.h

```diff
@@ -349,6 +348,155 @@ extern "C"
+    *           For the standard less-granular uploads to blob please use either
+    *           `IoTHubDeviceClient_UploadToBlobAsync` or `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.
+    *
+    * @param    iotHubClientHandle      The handle created by a call to the create function.
+    * @param    destinationFileName     Name of the file in blob storage.
+    * @param    uploadCorrelationId     Variable where to store the correlation-id of the new upload.
+    * @param    azureBlobSasUri         Variable where to store the Azure Storage SAS URI for the new upload.
+    *
+    * @warning  This is a synchronous/blocking function.
+    *           `uploadCorrelationId` and `azureBlobSasUri` must be freed by the calling application
+    *           after the blob upload process is done (e.g., after calling `IoTHubDeviceClient_NotifyUploadCompletion`).
+    *
+    * @return   An IOTHUB_CLIENT_RESULT value indicating the success or failure of the API call.
+    */
+    MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubDeviceClient_InitializeUpload, IOTHUB_DEVICE_CLIENT_HANDLE, iotHubClientHandle, const char*, destinationFileName, char**, uploadCorrelationId, char**, azureBlobSasUri);
+
+     /**
+    * @brief    This API creates a context for a new blob upload to Azure Storage.
+    * @remark   It is part of a set of functions for more granular control over Azure IoT-based blob uploads.
+    *           This function is expected to be used along with:
+    *           `IoTHubDeviceClient_CreateUploadContext`
+    *           `IoTHubDeviceClient_AzureStoragePutBlock`
+    *           `IoTHubDeviceClient_AzureStoragePutBlockList`
+    *           `IoTHubDeviceClient_DestroyUploadContext`
+    *           `IoTHubDeviceClient_InitializeUpload`
+    *           `IoTHubDeviceClient_NotifyUploadCompletion`
+    *           For the standard less-granular uploads to blob please use either
+    *           `IoTHubDeviceClient_UploadToBlobAsync` or `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.
+    *
+    * @param    iotHubClientHandle      The handle created by a call to the create function.
+    * @param    azureBlobSasUri         The Azure Storage Blob SAS uri obtained with `IoTHubDeviceClient_InitializeUpload`.
+    *
+    * @warning  This is a synchronous/blocking function.
+    *
+    * @return   A `IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE` on success or NULL if the function fails.
+    */
+    MOCKABLE_FUNCTION(, IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE, IoTHubDeviceClient_CreateUploadContext, IOTHUB_DEVICE_CLIENT_HANDLE, iotHubClientHandle, const char*, azureBlobSasUri);
+
+    /**
+    * @brief    This API uploads a single blob block to Azure Storage (performs a PUT BLOCK operation).
+    * @remark   It is part of a set of functions for more granular control over Azure IoT-based blob uploads.
+    *           This function is expected to be used along with:
+    *           `IoTHubDeviceClient_CreateUploadContext`
```

Comment:
If these are related, how about a more consistent naming convention like:

    *           `IoTHubDeviceClient_AzureStorageCreateUploadContext`
    *           `IoTHubDeviceClient_AzureStoragePutBlock`
    *           `IoTHubDeviceClient_AzureStoragePutBlockList`
    *           `IoTHubDeviceClient_AzureStorageDestroyUploadContext`
    *           `IoTHubDeviceClient_AzureStorageInitializeUpload`
    *           `IoTHubDeviceClient_AzureStorageNotifyUploadCompletion`

Reply:
It makes sense except for 
*           `IoTHubDeviceClient_AzureStorageInitializeUpload`
*           `IoTHubDeviceClient_AzureStorageNotifyUploadCompletion`
because these are not operations that target Azure Storage. They interact with Azure IoT Hub

Reply:
Ok, renamed them to
IoTHubDeviceClient_AzureStorageInitializeBlobUpload
IoTHubDeviceClient_AzureStorageNotifyBlobUploadCompletion